core/schema/file.go

```diff
@@ -25,6 +25,12 @@ func (s *fileSchema) Install() {
 			Doc(`Retrieves the size of the file, in bytes.`),
 		dagql.Func("name", s.name).
 			Doc(`Retrieves the name of the file.`),
+		dagql.Func("digest", s.digest).
```

Comment:
One thing worth consideration is if the digest should include the file metadata in addition to the file contents (i.e. owner, permissions, timestamps, etc.). The file metadata is part of its definition, impacts caching, etc. so depending on the use case it may or may not be what the user wants.

We can just add an optional argument for whether or not metadata is included of course, so then the question is just what the default behavior is (metadata included or not).

Personally, I have a slight preference for including metadata in the digest by default, but curious if others have thoughts.

Not a blocker for me, just raising the question.

Reply:
I think I'd expect the default to just be a content digest - I can imagine a lot of the cases of wanting a digest of the file to just get the contents - especially as that's what lots of users would be use to using `sha256sum` or similar command line tools.

Having this as an option definitely makes sense though (but - there's no authoritative way to hash file metadata in a consistent way - it feels a bit weird to have something non-standard like this, but I don't have any better suggestions).

Reply:
Is it possible with the buildkit cache digest implementation? I didn't see the options to do so

Reply:
> Having this as an option definitely makes sense though (although, there's no authoritative way to hash file metadata in a consistent way - it feels a bit weird to have something non-standard like this, but I don't have any better suggestions).

The closest thing to a standard is the tar format, which includes all file metadata. That's how container layer digests are computed, so I suppose it's part of the OCI standard in that respect.

That's also what the `contenthash` package ends up doing (without creating an actual tar file on disk): https://github.com/sipsma/buildkit/blob/699623575173eedf09b43b797bd74439c86a3071/cache/contenthash/filehash.go#L47-L47

> Is it possible with the buildkit cache digest implementation? I didn't see the options to do so

The `Checksum` implementation you switched to does include the file metadata in addition to the file contents. You're right it doesn't currently have an option to disable inclusion of metadata.

Like I said before, I'm personally fine with including the metadata by default and adding an option down the line to enable digesting only based off the contents. There's a few approaches for doing that (not needed now IMO, just noting for the future):
* Upstream changes to the buildkit package to support that
* When that "contents-only" option is enabled, reset all the metadata on the file so it's consistent and then get the hash from `Checksum`
* Just have a separate codepath when "contents-only" is enabled where we just read the contents and hash those directly rather than use the `Checksum` function

Reply:
BTW, I do see your point Justin about contents-only being the expectation from using `sha256sum` and similar.

I feel torn overall still because in the rest of Dagger the file metadata does impact the "checksum" of a file (and caching etc.) so including it by default feels aligned with the rest of our behavior. Really splitting hairs here though.

So still fine with including it by default, but not strongly opinionated if there's any more thoughts on this topic.

Reply:
I added the `excludeMetadata` options with the separate codepath :D

Reply:
If I had to pick a side I think I'd argue for alignment with `sha56sum` being less surprising, and if we want a digest that includes the metadata and anything else that influences caching, maybe that should be something like `cacheKey`.

Going back to the original request (#7709) it sounds like one goal is to see if two files are "the same" - which I think most people would take to mean same contents, metadata aside, the same way `sha256sum` etc. works.

But, yeah, low conviction and I'm coming in late so I'm good with shipping in either form

Reply:
Since @TomChv added support for both codepaths, it's trivial to switch the default behavior to not including metadata. Seems like more are leaning towards switching to that default, so we have time before the release to switch it.